src/com/t_oster/liblasercut/drivers/K40NanoVectorProperty.java

```diff
@@ -0,0 +1,166 @@
+/**
+ * This file is part of LibLaserCut.
+ * Copyright (C) 2011 - 2014 Thomas Oster <mail@thomas-oster.de>
+ *
+ * LibLaserCut is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * LibLaserCut is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with LibLaserCut. If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+package com.t_oster.liblasercut.drivers;
+
+import com.t_oster.liblasercut.LaserProperty;
+
+/**
+ *
+ * @author Tat
+ */
+public class K40NanoVectorProperty implements LaserProperty
+{
+  
+  static final String VAR_MM_PER_SECOND = "mm per second";
+  static final String VAR_D_RATIO = "diagonal speed adjustment";

```

Comment:
I'm not sure if this is a per-material property. If it only depends on the speed and not on the material itself, maybe it should rather be a machine property. Either as one constant, or two lists: a list of speeds and a corresponding list of diagonal factors.
For an example of how to use lists in the settings editor, see the LTT driver:
getProperty():
https://github.com/t-oster/LibLaserCut/blob/11bcce9051f8905103a0995923882a45a62071c6/src/com/t_oster/liblasercut/drivers/LaserToolsTechnicsCutter.java#L1823-L1830
setProperty():
https://github.com/t-oster/LibLaserCut/blob/11bcce9051f8905103a0995923882a45a62071c6/src/com/t_oster/liblasercut/drivers/LaserToolsTechnicsCutter.java#L1912-L1932

Reply:
Basically when both stepper motors trigger at the same time it takes a diagonal step which is longer than a purely ortogonal step. And the additional length is `sqrt(2) - 1` this means that the speed of the machine  ultimately depends, in part, on the angle it's drawing a line at. There's some other functional distances to expected idealized differences because since it takes tiny steps the the line drawing algorithm at the lowest level will result in some error giving your overall speed between 100% and 141% of the tickrate of the stepper motors. To try and minimize error there's a factor of addition ticks to wait between stepper motor pulses that is appended to the M2 and B2 boards. This tends to be 26% was their "auto" speed in the chinese software. But, this factor can be set to whatever, though really it shouldn't ever actually be changed. But, again, it's an actual factor the machine actually uses so I offered it up.

 If you are only cutting purely diagonal and purely orthogonal direction lines then it's better to use 0.41 there, as it will make the cut depths identical. 

![chinese-sqrt2](https://user-images.githubusercontent.com/3302478/59571099-2b53a500-9056-11e9-87b6-dca0df95a4da.png)

Reply:
It's built into the machine. And does matter depending on the type of cut you are making. You could possibly want one value for one type of cut and a different one for another. I can't see it being popular, or even common, or anything but cryptic and ignored but the speedcode and diagonal adjustments are two different elements. I'm strongly lead to believe that this is where that property should end up. If it were machine based it you couldn't change it from class of cut to another. When there could be cause to cut things at 55mm/s with a diagonal speed addition of 0.26 because you're cutting a bezier curve and cutting at 55mm/s with a diagonal speed addition of 0.41 because all your cuts are purely orthogonal and diagonal. 

It's cryptic as heck even in the chinese software that gives you a random ass percent and no explanation of what it was, but since it modified the speed code it was analyzed, reverse engineered, and reproven. My research on the subject even made it into the K40 Whisperer since he hadn't reverse engineered the speed codes very well. And research might be needed to find values for circles etc, as these might actually have points of minimum error at different values.